pkg/controller/masteruserrecord/masteruserrecord_controller_test.go

```diff
@@ -0,0 +1,200 @@
+	// given
+	logf.SetLogger(logf.ZapLogger(true))
+	s := apiScheme(t)
+	mur := newMasterUserRecord("john")
+	memberClient := fake.NewFakeClientWithScheme(s)
+	hostClient := fake.NewFakeClientWithScheme(s, mur)
+	cntrl := newController(hostClient, memberClient, s, newGetMemberCluster(true, v1.ConditionTrue))
+
+	// when
+	_, err := cntrl.Reconcile(newMurRequest(mur))
+
+	// then
+	require.NoError(t, err)
+	ua := &toolchainv1alpha1.UserAccount{}
+	err = memberClient.Get(context.TODO(), namespacedName(memberOperatorNs, "john"), ua)
+	require.NoError(t, err)
+	assert.EqualValues(t, mur.Spec.UserAccounts[0].Spec, ua.Spec)
+	assertMurCondition(t, hostClient, toBeNotReady(provisioningReason, ""))
+}
+
+func TestCreateUserAccountFailed(t *testing.T) {
+	// given
+	logf.SetLogger(logf.ZapLogger(true))
+	s := apiScheme(t)
+	mur := newMasterUserRecord("john")
+	memberClient := fake.NewFakeClientWithScheme(s)
+	hostClient := fake.NewFakeClientWithScheme(s, mur)
+
+	t.Run("when member cluster does not exist", func(t *testing.T) {
+		// given
+		cntrl := newController(hostClient, memberClient, s, newGetMemberCluster(false, v1.ConditionTrue))
+
+		// when
+		_, err := cntrl.Reconcile(newMurRequest(mur))
+
+		// then
+		require.Error(t, err)
+		msg := "the fedCluster member-cluster not found in the registry"
+		assert.Contains(t, err.Error(), msg)
+		assertUaNotFound(t, memberClient)
+		assertMurCondition(t, hostClient, toBeNotReady(clusterNotReady, msg))
+	})
+
+	t.Run("when member cluster does not exist", func(t *testing.T) {
+		// given
+		cntrl := newController(hostClient, memberClient, s, newGetMemberCluster(true, v1.ConditionFalse))
+
+		// when
+		_, err := cntrl.Reconcile(newMurRequest(mur))
+
+		// then
+		require.Error(t, err)
+		msg := "the fedCluster member-cluster is not ready"
+		assert.Contains(t, err.Error(), msg)
+		assertUaNotFound(t, memberClient)
+		assertMurCondition(t, hostClient, toBeNotReady(clusterNotReady, msg))
+	})
+}
+
+func assertMurCondition(t *testing.T, hostClient client.Client, expected toolchainv1alpha1.Condition) {
+	record := &toolchainv1alpha1.MasterUserRecord{}
+	err := hostClient.Get(context.TODO(), namespacedName(hostOperatorNs, "john"), record)
+	require.NoError(t, err)
+	test.AssertConditionsMatch(t, record.Status.Conditions, expected)
+}
+
+func assertUaNotFound(t *testing.T, memberClient client.Client) {
+	ua := &toolchainv1alpha1.UserAccount{}
+	err := memberClient.Get(context.TODO(), namespacedName(memberOperatorNs, "john"), ua)
+	require.Error(t, err)
+	assert.IsType(t, metav1.StatusReasonNotFound, errors.ReasonForError(err))
+}
+
+func newMasterUserRecord(userName string) *toolchainv1alpha1.MasterUserRecord {
+	mur := &toolchainv1alpha1.MasterUserRecord{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      userName,
+			Namespace: hostOperatorNs,
+		},
+		Spec: toolchainv1alpha1.MasterUserRecordSpec{
+			UserID: "12345abcdef",
+			UserAccounts: []toolchainv1alpha1.UserAccountEmbedded{{
```

Comment:
Ideally we would want to test multiple UserAccounts in one MUR but it would be pretty tricky to do in the current e2e setup (one host and one member)...

Reply:
Good point. I started working with multiple UserAccounts in the following PR, but it makes sense to do it here. 
As for e2e setup, yes it would be very tricky, but I'll cover it in the unit test

Reply:
added the test here: 13cea3e see https://github.com/codeready-toolchain/host-operator/pull/41/commits/13cea3e32a7e039d95c6ac284eada3b6429066dd#diff-93c8b7a9cc0875ddd6d3002d3d0b4253R51-R75
to make the code (and my life) easier, I also introduced some test doubles and assertions (that were originally introduced in the PR #23) as part of this commit. Please take a look at it, thanks and sorry for bringing new additional changes, but the code would look horrible without doing it :-)